### PR TITLE
Add preview for .ipynb files in pl-file-preview

### PR DESCRIPTION
**Pull request by kurmasz**
Added a workspace preview for .ipynb files in the pl-file-preview using notebookjs.
Notebookjs uses DOMpurify to sanitize the rendered markdown.
Future work could include:

Better styling of the markdown
Adding syntax highlighting
Adding a button to hide the Markdown blocks (assuming most instructors primarily care about the code blocks).
https://github.com/jsvine/notebookjs

### CODE REVIEW

File: yarn.lock
Comment (nwalters512) [inline yarn.lock:11447]: It's unfortunate that this pulls in a version of `jsdom` that's effectively unused; this leaves us with an extra 4MB of dependencies to ship around. I'm wondering if we could resolve this with Yarn resolutions: https://yarnpkg.com/configuration/manifest#resolutions

I'm happy to try out this change and push it to your branch if that's OK with you.

Comment (kurmasz) [reply yarn.lock:11447]: I'm good with that.
I'll fix the first two issues Tuesday or Wednesday.

File: apps/prairielearn/elements/pl-file-preview/pl-file-preview.js
Comment (nwalters512) [inline apps/prairielearn/elements/pl-file-preview/pl-file-preview.js:123]: I'd recommend moving the `await blob.text()` before you import the libraries to ensure that we consume the response body even if the dynamic imports fail for some reason.

Comment (kurmasz) [reply apps/prairielearn/elements/pl-file-preview/pl-file-preview.js:123]: Would it work just to add it to `Promise.all()`?

File: apps/prairielearn/elements/pl-file-preview/pl-file-preview.js
Comment (kurmasz) [reply apps/prairielearn/elements/pl-file-preview/pl-file-preview.js:120]: What exactly does the `await` do?  Does it cause any exceptions to be reported? Doesn't that happen anyway?

File: .yarnrc.yml
Comment (kurmasz) [reply .yarnrc.yml:1]: Would it work to delete my `yarn.lock` and re-run `make deps`?

Comment (nwalters512) [reply .yarnrc.yml]: `yarn.lock` is managed by `yarn` but it's not deterministic. Specifically it looks like this branch picked up our transition to Yarn 4 sometime during its lifetime, and Yarn's automatic upgrade machinery didn't do quite what we wanted it to.

You should run `git checkout master -- yarn.lock .yarnrc.yml` to restore the version on `master`, then run `yarn` to make only the necessary changes.

File: apps/prairielearn/elements/pl-file-preview/pl-file-preview.js
Comment (nwalters512) [reply apps/prairielearn/elements/pl-file-preview/pl-file-preview.js]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await

File: apps/prairielearn/elements/pl-file-preview/pl-file-preview.css
Comment (nwalters512) [reply apps/prairielearn/elements/pl-file-preview/pl-file-preview.css:34]: Ah, I didn't realize it was totally unstyled. I'll take a look at these and see if they should be changed at all.

File: apps/prairielearn/elements/pl-file-preview/pl-file-preview.js
Comment (nwalters512) [reply apps/prairielearn/elements/pl-file-preview/pl-file-preview.js:119]: You could, but it's conceptually clearer if there's just a single `const text = await blob.text()` that's used in both branches here.

**Comment by kurmasz** [comment]
@nwalters512 Ok, I think I got everything cleaned up and de-linted.

File: apps/prairielearn/elements/pl-file-preview/pl-file-preview.js
Comment (jonatanschroeder) [inline apps/prairielearn/elements/pl-file-preview/pl-file-preview.js:117]: I have a bit of a reservation of introducing yet another markdown library. We're currently using showdown in pl-file-editor (markdown preview), pl-rich-text-editor and pl-xss-safe, and unified in server-side conversion. Can we settle on one of those?

File: apps/prairielearn/elements/pl-file-preview/pl-file-preview.js
Comment (jonatanschroeder) [inline apps/prairielearn/elements/pl-file-preview/pl-file-preview.js:122]: For other elements that perform sanitization we also include the `SANITIZE_NAMED_PROPS` option, to avoid issues with conflicts with other elements.
```suggestion
                    nb.sanitizer = (code) => DOMPurify.sanitize(code, {SANITIZE_NAMED_PROPS: true});
```

File: apps/prairielearn/elements/pl-file-preview/pl-file-preview.js
Comment (kurmasz) [reply apps/prairielearn/elements/pl-file-preview/pl-file-preview.js:117]: I get the following error when I try to dynamically load showdown:

`TypeError: undefined is not an object (evaluating 'this.showdown=x') — showdown.js:5153`

* Does this look familiar (i.e., might there be a quick, easy fix)?
* I could just statically load it; but, I don't think @nwalters512 would want this library loaded unless the file being previewed is .ipynb.

Comment (jonatanschroeder) [reply apps/prairielearn/elements/pl-file-preview/pl-file-preview.js:117]: To clarify: I don't want to block this use, particularly considering #6480, I'm just questioning if the use of a new library for functionality that is already in place. If there is no easy alternative please proceed.

Comment (kurmasz) [reply apps/prairielearn/elements/pl-file-preview/pl-file-preview.js:117]: I understand. I agree that fewer libraries is better, it just a question of the best tradeoff: Fewer libraries, dynamic loading, or (potentially) waiting for a PR in showdown (assuming the error is their bug, not mine). I'm fine with whatever you prefer.

File: apps/prairielearn/elements/pl-file-preview/pl-file-preview.js
Comment (jonatanschroeder) [reply apps/prairielearn/elements/pl-file-preview/pl-file-preview.js:122]: Did you push?

Comment (kurmasz) [reply apps/prairielearn/elements/pl-file-preview/pl-file-preview.js:122]: No. I'm taking a few minutes to see if the showdown problem is a quick fix. (I should be working on getting my classes ready, but the temptation to play with code was too great :( ).

Comment (jonatanschroeder) [reply apps/prairielearn/elements/pl-file-preview/pl-file-preview.js:122]: > I should be working on getting my classes ready, but the temptation to play with code was too great :(

Preach...

Comment (kurmasz) [reply apps/prairielearn/elements/pl-file-preview/pl-file-preview.js:122]: It turns out that the showdown problem is the same `this` vs. `this || globalThis` problem that `notebookjs` had. I'm going to prepare an MVE and PR for showdown, but then revert this PR to using Marked.  I'll make a new PR if/when showdown updates.  (I don't want to risk delaying this feature another three months.)  I'll let you know when I've pushed.

**Comment by kurmasz** [comment]
Ok, third time's a charm, right?

File: apps/prairielearn/elements/pl-file-preview/pl-file-preview.js
Comment (nwalters512) [reply apps/prairielearn/elements/pl-file-preview/pl-file-preview.js]: For posterity, here's the PR into showdown: https://github.com/showdownjs/showdown/pull/1017

That project hasn't seen a merge in years, so I'm not optimistic about it being fixed 😢

File: apps/prairielearn/elements/pl-file-preview/pl-file-preview.js
Comment (nwalters512) [reply apps/prairielearn/elements/pl-file-preview/pl-file-preview.js:117]: @jonatanschroeder could we move in the other direction and switch everything else to use `marked`? I'm not sure what compatibility looks like, but it does support dynamic imports (good for lazy-loading when needed), and it's less than half the size ([10.6kB](https://bundlephobia.com/package/marked@14.0.0) vs [23.lkB](https://bundlephobia.com/package/showdown@2.1.0)).

This doesn't need to block this PR, of course.

**Comment by nwalters512** [comment]
@kurmasz if there aren't any merge conflicts, there's no need to constantly merge with master. CI will automatically test the merge result before this lands. I'm taking a final pass over this now.

File: apps/prairielearn/elements/pl-file-preview/pl-file-preview.js
Comment (nwalters512) [inline apps/prairielearn/elements/pl-file-preview/pl-file-preview.js:142]: Addition: we'll typeset any math with `MathJax`.

File: apps/prairielearn/elements/pl-file-preview/pl-file-preview.js
Comment (nwalters512) [inline apps/prairielearn/elements/pl-file-preview/pl-file-preview.js:161]: This is opinionated, but I think it's the right thing to do. A 200px-tall preview of a notebook is almost never going to be what someone wants.

File: apps/prairielearn/elements/pl-file-preview/pl-file-preview.mustache
Comment (nwalters512) [inline apps/prairielearn/elements/pl-file-preview/pl-file-preview.mustache:8]: I changed this so that any lists in a notebook preview would render at the correct depth.

File: apps/prairielearn/elements/pl-file-preview/pl-file-preview.mustache
Comment (nwalters512) [inline apps/prairielearn/elements/pl-file-preview/pl-file-preview.mustache:30]: This was changed in the interest of separating the expansion behavior from whether we're displaying a plaintext or rendered notebook preview.

**Review by nwalters512 — APPROVED** [review]
I tweaked things a bit in 72a7230. I'll leave a few review comments explaining what I changed, but the main difference was to not render the notebook preview in a `<pre>`, but rather in a standard div. This looks nicer IMO, and means that we only need a single custom style (just something to give any `pre` elements a background to set them apart from plain text).

I'm quite happy with where this is at now. @kurmasz since I pushed to your branch, I'd welcome any comments you have about my changes, and I'd encourage you to test the updated styles with some notebooks of your own. Let me know if everything looks good to you and I'll get this merged!

**Comment by kurmasz** [comment]
@nwalters512 all you changes look good to be; but, I can't run them locally. I can no longer launch workspaces. (After pulling the branch, I ran`rm -rf node_modules` and `make deps`.)

File: apps/prairielearn/elements/pl-file-preview/pl-file-preview.js
Comment (jonatanschroeder) [reply apps/prairielearn/elements/pl-file-preview/pl-file-preview.js:123]: > @jonatanschroeder could we move in the other direction and switch everything else to use `marked`? I'm not sure what compatibility looks like, but it does support dynamic imports (good for lazy-loading when needed), and it's less than half the size ([10.6kB](https://bundlephobia.com/package/marked@14.0.0) vs [23.lkB](https://bundlephobia.com/package/showdown@2.1.0)).
> 
> This doesn't need to block this PR, of course.

Ideally I'd prefer to add bundling to @prairielearn/markdown, so we have consistency between server-side and client-side conversion, but marked may be an alternative if that package ends up being too big or if it takes too long to get this rolling. 

That said, one option we could also consider is to switch everything to marked, including server-side. Marked seems to have [the extensibility we need](https://marked.js.org/using_pro) to convert, e.g., pre to pl-code. It also has native support for [inline parsing](https://marked.js.org/using_advanced#inline), which is used in the context of rubrics, and that we currently handle in a somewhat hacky way. It doesn't handle sanitizing, but we can combine it with DOMPurify.

The only issue we might have with marked is that marked does not support HTML to markdown conversion, which is [used in pl-rich-text-editor element](https://github.com/PrairieLearn/PrairieLearn/blob/ee607c24e6461abe4abd4edb6a017292f9461da1/apps/prairielearn/elements/pl-rich-text-editor/pl-rich-text-editor.js#L123). This is a point where unified [could be used for this purpose](https://github.com/rehypejs/rehype-remark). That said, to be honest, with the problems we have had with showdown in this respect (see #5986, #10026), I would not be opposed to splitting this to use a different package altogether, potentially even in the Python side of pl-rich-text-editor itself (i.e., the parse function) instead of client-side.

Comment (jonatanschroeder) [reply apps/prairielearn/elements/pl-file-preview/pl-file-preview.js:123]: Added this discussion to #6480. And to be clear, please proceed with marked in this PR.

**Comment by nwalters512** [comment]
I'm going to land this as-is. We can consider styling tweaks in the future.

**Comment by kurmasz** [comment]
That makes sense to me.  Thanks for all your help.

**Pull request merged**
